### PR TITLE
WIP: 235 contract validation fixes

### PR DESCRIPTION
**Pull request by Divinare**
Closes https://github.com/HSLdevcom/bultti/issues/235

* Didnd't validate "removing unselecting contract being used for procurementUnit" - As admin access rights are required in the modifyContract resolver, I think we can trust frontend to prevent this
* Merging this PR will make existing preInspection that are !== draft related checks not work because they don't have a procurementUnitContractMap initialized.
  * Remove/modify contract isn't prevented
  * Contract page, procurementUnit unselecting is not disabled
  * I had to leave some old code to prevent this kind of situation to cause errors, left a comment above it
     // TODO: remove the code below when all of the existing preInspections have a procurementUnitContractMap
* Edit contract's rules file / remove contract: added guarding to prevent action if something is found from procurementUnitContractMap
  * removed old remove validation which used contract start dates
  
ContractManager's procurementUnit cache is marked with TODO: It can be fixed somehow yes, but I think we should consider refactoring models a bit (split ExecutionRequirement into PlannedAreaRequirement and PlannedProcurementUnitRequirement). That task would cause changes to ContractManager anyway...

### CODE REVIEW

File: src/inspection/InspectionResolver.ts [hunk 1]
```diff
@@ -607,7 +610,41 @@ export class InspectionResolver implements ResolverInterface<Inspection> {
...
+        let currentUnitContract = contractCandidates?.[0]
+        if (currentUnitContract) {
+          unitContractMap[unit!.id] = currentUnitContract.id
+        } else {
+          if (unit) {
+            throw new ApolloError(`Contract not found for procurementUnit: ${unit.id}`)
+          }
+        }
+      }
+      inspection.procurementUnitContractMap = unitContractMap
```
Comment (danieldunderfelt) [inline src/inspection/InspectionResolver.ts:642]: This is not only for pre-inspections. Use the `getProcurementUnitsByInspection` function to get the units for any inspection and take the contracts from there.

**Review by danieldunderfelt — CHANGES_REQUESTED** [review]
A good start, but not quite ready yet!

File: src/inspection/inspectionUtils.ts [hunk 2]
```diff
@@ -114,6 +116,75 @@ export function getBaseInspectionType(inspection: Inspection) {
...
+  //  Example 3 of Condition A) - A is the only candidate
+  //  Contract A |--------------------------|
+  //  Contract B         |-------|
+  //  Inspection |--------------------|
+  //
+  //  Example 4 of Condition A) - There are no candidates
+  //  Contract A |-----------------|
+  //  Contract B              |-------|
+  //  Inspection     |--------------------|
+  let isMatchingConditionA = (contract: Contract): boolean => {
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/inspection/inspectionUtils.ts:157]: This is what I was thinking but here the variable names would get too long:
isContractStartDateWithinInspectionStartDateAndContractEndDateGreaterThanEqualInspectionEnDate
isInspectionFullyCoveredWithContract

More readable is just 2 condition variables. If you want to know exactly what the condition is, read comments or the code.

[the remark this answers is not in the document]
Comment (Divinare) [reply src/inspection/inspectionUtils.ts:157]: Well, one option would be to split that first into two conditions

isContractStartDateWithinInspectionDates
isContractEndDateGreaterOrEqualThanInspectionEndDate
and use them combined with && but i dont think it adds any value to readability

File: src/contract/ContractManager.ts [hunk 3]
```diff
@@ -73,25 +81,35 @@ export class ContractManager {
...
+    let contractId = this.procurementUnitContractMap[procurementUnitId]
+    let contractRepo = getRepository(Contract)
+    let contract = await contractRepo.findOne({
+      id: contractId,
+    })
+    if (contract) {
+      return contract
+    }
+
+    // TODO: remove the code below when all of the existing preInspections have a procurementUnitContractMap
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/contract/ContractManager.ts:99]: Hmm.... I think better would be to have procurementUnitsMap never empty. When an inspection is created as a draft, we initialize  procurementUnitsMap. When an inspection is made draft -> review, we update procurementUnitsMap

File: src/contract/ContractResolver.ts [hunk 4]
```diff
@@ -409,3 +431,36 @@ export class ContractResolver {
     }))
   }
 }
+
+const guardContractBeingUsedInPreInspection = async (
```
[the remark this answers is not in the document]
Comment (Divinare) [reply src/contract/ContractResolver.ts:435]: Okay, I think this is a code convention so I updated it here:
https://github.com/HSLdevcom/bultti-ui/pull/64/commits/8532bf85ca9164a90e232394cfabbb5eec42b1de

File: src/contract/ContractManager.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/contract/ContractManager.ts:99]: Let's do it that way, yes.

**Comment by Divinare** [comment]
The spec for implenting things related to this is the air, it seems not smart trying to iterate all of this when we aren't even sure of how things should go. This is why we should stop here (leave it in a stable state though!)

--> Left for future tasks:
1) block changing of dates with any non-draft inspections
  - maybe make startDate as uneditable
  - get minimum endDate for contract
2) tests to "get contract date candidates"

What was done here?
0) Review fixes done
1) Contract's cant overlap by clicking "Sopimuksessa mukana"
 - HOWEVER, if you edit contract dates, this can cause contracts to overlap. I think we need lock specbefore going any further with this.
2) You have a temprorary button to init inspection.procurementUnitContractMap

File: src/inspection/inspectionUtils.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (danieldunderfelt) [reply src/inspection/inspectionUtils.ts]: Well... the name doesn't need to include everything. But let's go with the current state for now.

**Pull request merged**